Admin/components/AdminSite/TwoFactorAuthentication.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+use RobThree\Auth\TwoFactorAuth;
+
+/**
+ * Authenticate 2FA Token
+ *
+ * @package   Admin
+ * @copyright 2022 silverorange
+ */
+class AdminAdminSiteTwoFactorAuthentication extends AdminPage
+{
+	// init phase
+	// {{{ protected function createLayout()
+
+	protected function createLayout()
+	{
+		return new AdminLoginLayout($this->app, AdminLoginTemplate::class);
+	}
+
+	// }}}
+	// {{{ protected function initInternal()
+
+	protected function initInternal()
+	{
+		parent::initInternal();
+
+		$this->ui->loadFromXML(__DIR__.'/two_factor_authentication.xml');
+
+		$form = $this->ui->getWidget('two_fa_form');
+		$form->action = 'AdminSite/TwoFactorAuthentication';
+
+		// remember where we came from
+		$form->addHiddenField('relocate_uri', $this->app->getUri());
+
+		$user = $this->app->session->user;
+		if ($user->is2FaAuthenticated() || !$user->two_fa_enabled) {
+			$this->app->relocate('./');
+		}
+	}
+
+	// }}}
+
+	// process phase
+	// {{{ protected function processInternal()
+
+	protected function processInternal()
+	{
+		parent::processInternal();
+
+		$form = $this->ui->getWidget('two_fa_form');
+		if ($form->isProcessed()) {
+			if (!$form->hasMessage()) {
+				$this->validate2Fa();
+			}
+
+			if (!$form->hasMessage()) {
+				$this->app->session->user->set2FaAuthenticated();
+
+				// go back where we came from
+				$uri = $form->getHiddenField('relocate_uri');
+				$this->app->relocate($uri);
+			}
+		}
+	}
+
+	// }}}
+	// {{{ protected function validate2Fa()
+
+	protected function validate2Fa()
+	{
+		// strip all non numeric characters like spaces and dashes that people
+		// might enter (e.g. Authy adds spaces for readability)
+		$token = preg_replace(
```

Comment:
this logic is copied in the profile page. It makes sense to move it to a separate class used in both places that wraps around `TwoFactorAuth`